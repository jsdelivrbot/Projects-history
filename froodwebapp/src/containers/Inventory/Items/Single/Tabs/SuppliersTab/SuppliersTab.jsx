import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Row, Col } from 'react-flexbox-grid';
import { connect } from 'react-redux';
import {
  itemSuppliersGetRequest,
  itemSuppliersSaveRequest,
  itemSuppliersUpdateRequest
} from 'redux-base/actions';
import { Table, Spin, Input } from 'antd';
import { ActionButton, ConnectedTopFormModal } from 'components';
import { table } from 'styles/common.scss';
import fields from './modalFields';
import columns from './suppliersTabHelpers';

const { Search } = Input;

const mapStateToProps = state => ({
  uoms: state.uom.data,
  suppliers: state.item.suppliers,
  loadingPage: state.item.loadingPage,
  needReloadSuppliers: state.item.needReloadSuppliers,
  itemInfo: state.item.itemInfo,
});

const mapDispatchToProps = {
  itemSuppliersGetRequest,
  itemSuppliersSaveRequest,
  itemSuppliersUpdateRequest
};

const newSkuLink = {
  sku: '',
  supplierId: '',
  supplierSku: '',
  supplierSkuName: '',
  uomId: '',
  minPurchaseQuantity: '',
  isDefault: true
};

export class SuppliersTab extends Component {
  constructor(props) {
    super(props);
    this.state = {
      searchValue: '',
      modalVisible: false,
      modalTitle: 'New Supplier Link',
      modalData: newSkuLink
    };
  }

  componentWillReceiveProps(nextProps) {
    // when we edit or create new Link to reload vendors
    if (nextProps.needReloadSuppliers) {
      this.setState({
        modalData: undefined,
        modalVisible: !this.state.modalVisible
      }, () => this.props.itemSuppliersGetRequest({
        id: this.props.itemId
      }));
    }
  }

  handleSearchChange = (e) => {
    this.setState({
      searchValue: e.target.value
    });
  }

  handleEdit = (e) => {
    this.setState({
      modalTitle: 'Edit Supplier Link',
      modalVisible: true,
      modalData: this.props.suppliers.find(item => item.id === Number(e.target.id))
    });
  }

  handleToggleModal = () => {
    this.setState({
      modalTitle: 'New Supplier Link',
      modalVisible: !this.state.modalVisible,
      modalData: newSkuLink
    });
  }

  handleSave = (formData) => {
    const {
      supplierId,
      sku,
      supplierSku,
      supplierSkuName,
      uomId,
      minPurchaseQuantity,
      isDefault
    } = formData;

    if (this.state.modalData.id) {
      this.props.itemSuppliersUpdateRequest({
        id: this.props.itemId,
        variantId: this.state.variantId,
        payload: {
          id: this.state.modalData.id,
          sku,
          supplierSku,
          supplierSkuName,
          uomId,
          minPurchaseQuantity,
          isDefault,
          supplierId: supplierId.id
        }
      });
    } else {
      this.props.itemSuppliersSaveRequest({
        id: this.props.itemId,
        variantId: this.state.variantId,
        payload: {
          sku,
          supplierSku,
          supplierSkuName,
          uomId,
          minPurchaseQuantity,
          isDefault,
          supplierId: supplierId.id
        }
      });
    }
  }

  render() {
    const {
      modalData,
      modalTitle,
      modalVisible,
      searchValue
    } = this.state;

    const {
      loadingPage,
      suppliers,
      uoms,
      itemInfo
    } = this.props;

    const data = suppliers.filter(supplier =>
      supplier && supplier.sku.toLowerCase().includes(searchValue.toLowerCase())
    );

    const skus = itemInfo.variants
      ? itemInfo.variants.map(variant => ({ key: variant.id, value: variant.sku }))
      : ([{ key: itemInfo.id, value: itemInfo.sku }]);

    return (
      <div>
        <ConnectedTopFormModal
          loading={ loadingPage }
          title={ modalTitle }
          handleSave={ this.handleSave }
          handleToggleModal={ this.handleToggleModal }
          visible={ modalVisible }
          initialValues={ modalData }
          fields={ fields(uoms, skus) }
          buttonVisible={ false }
        />
        <Spin spinning={ loadingPage }>
          <Row end="lg" style={ { marginBottom: '1rem' } }>
            <Col lg={ 4 }>
              <Search
                placeholder="Type to search SKU/Variant"
                onChange={ this.handleSearchChange }
              />
            </Col>
            <Col lg>
              <ActionButton
                style={ { float: 'right' } }
                onClick={ this.handleToggleModal }
              >
                New Link
              </ActionButton>
            </Col>
          </Row>
          <Row>
            <Col sm md lg>
              <Table
                className={ table }
                columns={ columns(this.handleEdit) }
                dataSource={ data }
                size="small"
                pagination={ false }
                rowKey="id"
              />
            </Col>
          </Row>
        </Spin>
      </div>
    );
  }
}

SuppliersTab.propTypes = {
  // trigger
  loadingPage: PropTypes.bool.isRequired,
  needReloadSuppliers: PropTypes.bool.isRequired,
  // data
  suppliers: PropTypes.array.isRequired,
  uoms: PropTypes.array.isRequired,
  itemInfo: PropTypes.object.isRequired,
  itemId: PropTypes.string.isRequired,
  // redux-base
  itemSuppliersGetRequest: PropTypes.func.isRequired,
  itemSuppliersSaveRequest: PropTypes.func.isRequired,
  itemSuppliersUpdateRequest: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(SuppliersTab);

