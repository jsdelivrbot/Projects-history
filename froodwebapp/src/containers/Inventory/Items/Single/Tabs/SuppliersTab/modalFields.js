import { getMenuItems } from 'utils';

export default (uoms, skus) => ([{
  header: 'SKU/Variant',
  type: 'select',
  name: 'sku',
  editableForExistingItem: false,
  editableForNewItem: true,
  menuItems: skus
}, {
  header: 'Choose Supplier Name or Add new',
  type: 'supplierAutoComplete',
  name: 'supplierId',
  editableForExistingItem: false,
  editableForNewItem: true,
}, {
  header: 'Supplier SKU',
  type: 'text',
  name: 'supplierSku',
  editableForExistingItem: false,
  editableForNewItem: true,
}, {
  header: 'Suplier Product Name',
  type: 'text',
  name: 'supplierSkuName',
  editableForExistingItem: false,
  editableForNewItem: true,
}, {
  header: 'UOM',
  type: 'select',
  name: 'uomId',
  editableForExistingItem: false,
  editableForNewItem: true,
  menuItems: getMenuItems(uoms)
}, {
  header: 'Min Purchase Quantity',
  type: 'number',
  name: 'minPurchaseQuantity',
  editableForExistingItem: false,
  editableForNewItem: true,
}, {
  text: 'Make this suppplier default for SKU',
  type: 'checkbox',
  name: 'isDefault',
  editableForExistingItem: true,
  editableForNewItem: true,
}]);

