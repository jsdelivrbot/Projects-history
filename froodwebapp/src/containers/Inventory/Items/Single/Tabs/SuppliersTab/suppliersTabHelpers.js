import React from 'react';
import { TableActionButtons } from 'components';
import { Checkbox } from 'antd';

export default handleEdit => [{
  title: 'SKU',
  dataIndex: 'sku',
}, {
  title: 'SUPPLIER',
  dataIndex: 'supplierName',
}, {
  title: 'SUPPLIER SKU',
  dataIndex: 'supplierSku',
}, {
  title: 'SUPPLIER PRODUCT NAME',
  dataIndex: 'supplierSkuName',
}, {
  title: 'PURCHASE UOM',
  dataIndex: 'uomName',
}, {
  title: 'MIN PURCHASE QTY',
  dataIndex: 'minPurchaseQuantity',
}, {
  title: 'LAST SUPPLIED',
  dataIndex: 'lastSupplied',
}, {
  title: 'DEFAULT',
  dataIndex: 'isDefault',
  render: checked => (
    <Checkbox checked={ checked } disabled />
  )
}, {
  title: 'ACTION',
  render: (text, record) => (
    <TableActionButtons
      record={ record }
      handleEdit={ handleEdit }
    />
  )
}];

