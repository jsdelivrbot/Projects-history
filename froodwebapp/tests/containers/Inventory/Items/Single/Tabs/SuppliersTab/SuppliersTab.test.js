import { SuppliersTab } from 'containers/Inventory/Items/Single/Tabs/SuppliersTab/SuppliersTab';
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

const { describe, expect, it } = global;

describe('SuppliersTab', () => {
  const props = {
    loadingPage: false,
    needReloadSuppliers: false,
    suppliers: [{
      sku: 'sku',
      id: '1'
    }],
    uoms: [],
    itemInfo: {
      variants: [{
        id: '1',
        sku: 'sku'
      }]
    },
    itemId: '1',
    itemSuppliersGetRequest: jest.fn(),
    itemSuppliersSaveRequest: jest.fn(),
    itemSuppliersUpdateRequest: jest.fn(),
  };

  it('renders the SuppliersTab component', () => {
    const component = shallow(<SuppliersTab { ...props } />);

    expect(component.length).toEqual(1);
  });

  it('capturing Snapshot of SuppliersTab', () => {
    const component = shallow(<SuppliersTab { ...props } />);
    const view = toJson(component);
    const actual = expect(view);

    actual.toMatchSnapshot();
  });
});
