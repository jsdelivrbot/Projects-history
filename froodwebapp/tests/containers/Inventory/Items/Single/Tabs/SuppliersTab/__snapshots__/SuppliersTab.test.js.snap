// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuppliersTab capturing Snapshot of SuppliersTab 1`] = `
<div>
  <ReduxForm
    buttonVisible={false}
    fields={
      Array [
        Object {
          "editableForExistingItem": false,
          "header": "SKU/Variant",
          "menuItems": Array [
            Object {
              "key": "1",
              "value": "sku",
            },
          ],
          "name": "sku",
          "type": "select",
        },
        Object {
          "editableForExistingItem": false,
          "header": "Choose Supplier Name or Add new",
          "name": "supplierId",
          "type": "supplierAutoComplete",
        },
        Object {
          "editableForExistingItem": false,
          "header": "Supplier SKU",
          "name": "supplierSku",
          "type": "text",
        },
        Object {
          "editableForExistingItem": false,
          "header": "Suplier Product Name",
          "name": "supplierSkuName",
          "type": "text",
        },
        Object {
          "editableForExistingItem": false,
          "header": "UOM",
          "menuItems": Array [],
          "name": "uomId",
          "type": "select",
        },
        Object {
          "editableForExistingItem": false,
          "header": "Min Purchase Quantity",
          "name": "minPurchaseQuantity",
          "type": "number",
        },
        Object {
          "editableForExistingItem": true,
          "name": "isDefault",
          "text": "Make this suppplier default for SKU",
          "type": "checkbox",
        },
      ]
    }
    handleSave={[Function]}
    handleToggleModal={[Function]}
    initialValues={
      Object {
        "id": "",
        "isDefault": true,
        "minPurchaseQuantity": "",
        "sku": "",
        "supplierId": "",
        "supplierSku": "",
        "supplierSkuName": "",
        "uomId": "",
      }
    }
    loading={false}
    title="New Supplier Link"
    visible={false}
  />
  <Spin
    prefixCls="ant-spin"
    size="default"
    spinning={false}
    wrapperClassName=""
  >
    <Row
      end="lg"
      style={
        Object {
          "marginBottom": "1rem",
        }
      }
    >
      <Col
        lg={4}
      >
        <Search
          onChange={[Function]}
          placeholder="Type to search SKU/Variant"
          prefixCls="ant-input-search"
        />
      </Col>
      <Col
        lg={true}
      >
        <ActionButton
          onClick={[Function]}
          style={
            Object {
              "float": "right",
            }
          }
        >
          New Link
        </ActionButton>
      </Col>
    </Row>
    <Row>
      <Col
        lg={true}
        md={true}
        sm={true}
      >
        <Table
          bordered={false}
          className="table"
          columns={
            Array [
              Object {
                "dataIndex": "sku",
                "title": "SKU",
              },
              Object {
                "dataIndex": "supplierName",
                "title": "SUPPLIER",
              },
              Object {
                "dataIndex": "supplierSku",
                "title": "SUPPLIER SKU",
              },
              Object {
                "dataIndex": "supplierSkuName",
                "title": "SUPPLIER PRODUCT NAME",
              },
              Object {
                "dataIndex": "uomName",
                "title": "PURCHASE UOM",
              },
              Object {
                "dataIndex": "minPurchaseQuantity",
                "title": "MIN PURCHASE QTY",
              },
              Object {
                "dataIndex": "lastSupplied",
                "title": "LAST SUPPLIED",
              },
              Object {
                "dataIndex": "isDefault",
                "render": [Function],
                "title": "DEFAULT",
              },
              Object {
                "render": [Function],
                "title": "ACTION",
              },
            ]
          }
          dataSource={
            Array [
              Object {
                "id": "1",
                "sku": "sku",
              },
            ]
          }
          indentSize={20}
          loading={false}
          locale={Object {}}
          pagination={false}
          prefixCls="ant-table"
          rowKey="id"
          rowSelection={null}
          showHeader={true}
          size="small"
          useFixedHeader={false}
        />
      </Col>
    </Row>
  </Spin>
</div>
`;
